plugin/plugin.go

```diff
@@ -141,11 +132,186 @@ func (p *AzurePlugin) OnUpdateSet(_ context.Context, req *pb.OnUpdateSetRequest)
+
+	responseStart := time.Now()
+	response := p.buildHostsResponse(vmToNetworkMap, resourceToSetMap)
+	fmt.Printf("Built response in %v (Total hosts: %d)\n",
+		time.Since(responseStart),
+		len(vmToNetworkMap))
+
+	return response, nil
+}
+
+func rotateCredFromCallback(ctx context.Context, catalog *hostcatalogs.HostCatalog) (*pb.HostCatalogPersisted, error) {
+	authzInfo, err := getAuthorizationInfo(catalog)
+	if err != nil {
+		return nil, fmt.Errorf("error getting auth config when creating catalog: %w", err)
+	}
+	newPass, err := rotateCredential(ctx, authzInfo)
+	if err != nil {
+		return nil, fmt.Errorf("error rotating credentials when creating catalog: %w", err)
+	}
+	if newPass == nil {
+		return nil, errors.New("new credential back from rotate is nil")
+	}
+	newSecrets, err := structpb.NewStruct(map[string]interface{}{
+		constSecretId:             *newPass.KeyId,
+		constSecretValue:          *newPass.SecretText,
+		constCredsLastRotatedTime: time.Now().Format(time.RFC3339Nano),
+	})
+	if err != nil {
+		return nil, fmt.Errorf("error formatting new credential data as struct")
+	}
+	return &pb.HostCatalogPersisted{
+		Secrets: newSecrets,
+	}, err
+}
+
+// -----------------------------------------------------------------------------
+// Response Building Functions
+// -----------------------------------------------------------------------------
+func (p *AzurePlugin) buildHostsResponse(vmToNetworkMap map[string]networkInfo,
+	resourceToSetMap map[string][]string) *pb.ListHostsResponse {
+
+	ret := &pb.ListHostsResponse{}
+
+	for resourceID, networkInfo := range vmToNetworkMap {
+		host := p.createHostFromResource(resourceID, networkInfo, resourceToSetMap)
+		if host != nil {
+			ret.Hosts = append(ret.Hosts, host)
+		}
+	}
+
+	return ret
+}
+
+func (p *AzurePlugin) createHostFromResource(resourceID string, networkInfo networkInfo,
+	resourceToSetMap map[string][]string) *pb.ListHostsResponseHost {
+
+	// Extract the resource type from the ID
+	resourceType := extractResourceType(resourceID)
+
+	switch resourceType {
+	case constVirtualMachineScaleSetsResource:
+		externalName, err := getExternalNameforVMSSInstance(resourceID)
+		if err != nil {
+			externalName = "" // Use empty name if there's an error
+		}
+
+		setID := getSetForVMSSInstance(resourceID)
+		return &pb.ListHostsResponseHost{
+			ExternalId:   resourceID,
+			ExternalName: externalName,
+			IpAddresses:  networkInfo.IpAddresses,
+			SetIds:       resourceToSetMap[setID],
+		}
+
+	case constVirtualMachinesResource:
+		_, externalName, err := splitId(resourceID, constMsComputeService, constVirtualMachinesResource)
+		if err != nil {
+			externalName = "" // Use empty name if there's an error
+		}
+
+		return &pb.ListHostsResponseHost{
+			ExternalId:   resourceID,
+			ExternalName: externalName,
+			IpAddresses:  networkInfo.IpAddresses,
+			SetIds:       resourceToSetMap[resourceID],
+		}
+
+	default:
+		return nil // Resource type not supported
```

Comment:
I think we should error defensively in this case, so we discover these unexpected cases earlier.